outlines/fsm/guide.py

```diff
@@ -274,3 +247,60 @@ def must_terminate_state(self, state: CFGState) -> bool:
     def copy(self) -> "CFGGuide":
         """Create a copy of the Guide."""
         return CFGGuide(self.cfg_string, self.tokenizer)
+
+
+class Write:
+    def __init__(self, text: str):
+        self.text = text
+
+
+class Generate:
+    def __init__(self, model, allowed_token_ids):
+        self.model = model
+        self.allowed_token_ids = allowed_token_ids
+
+
+def create_states_mapping(fsm, tokenizer):
+    states_mapping = {}
+    for state in fsm.states:
+        states_mapping[state] = {
+            "allowed_token_ids": list(fsm.transitions[state].keys()),
+            "is_final": fsm.is_final(state),
+        }
+    return states_mapping
+
+
+def create_states_mapping_from_fsm(fsm, tokenizer):
+    states_mapping = {}
+    for state in fsm.states:
+        states_mapping[state] = {
+            "allowed_token_ids": list(fsm.transitions[state].keys()),
+            "is_final": fsm.is_final(state),
+        }
+    return states_mapping
+
+
+class RegexGuide(Guide):
+    def __init__(self, regex, tokenizer):
+        self.regex = regex
+        self.tokenizer = tokenizer
+        self.eos_token_id = tokenizer.eos_token_id
+        self.fsm = build_regex_from_schema(regex)
+        self.states_mapping = create_states_mapping_from_fsm(self.fsm, tokenizer)
+        self.initial_state = 0
```

Comment:
I would suggest to start with the least amount of attributes possible and then iterate on what's actually needed.